### ISSUE TITLE
Project module type changed to ES module (commonjs removed)

### ISSUE THREAD

**Issue by clintonwoo** [blakeembrey/sql-template-tag#28]
Hey, nice project.

I was just wondering.. In `5.0.0` release (https://github.com/blakeembrey/sql-template-tag/releases/tag/v5.0.0) the project was changed to stop using CommonJS modules and instead use ES modules. Just wondering what the main motivation behind that was because technically doesn't this remove support for NodeJS versions that don't support ES modules natively (less than Node `14.8`-ish)? This can be a problem for bundlers targeting less recent versions of ES and many projects that depend on it.

Isn't the more common pattern to provide the library for both commonjs and esmodules using something like this:
```
"main": "dist/cjs/index.js",
"module": "dist/mjs/index.js",
"exports": {
    ".": {
        "import": "./dist/mjs/index.js",
        "require": "./dist/cjs/index.js"
    }
},
```

For example the problem is being explained in this article: https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html

**Comment by mindplay-dk** [comment]
Is this why the package won't load with `ts-node`?

@blakeembrey how do you use this package without `ts-node`? transpiling your whole project before run? 😕

**Comment by blakeembrey** [comment]
If you want old behavior, you can use the old version. ESM has been supported since node v12.17.0 without a CLI flag so all supported node.js versions can run this package. I'm well aware I can chose to support both, but I will not because it's a ton of overhead as a maintainer I don't want to deal with.

@mindplay-dk Yes, I transpile my projects or use `ts-node/esm`. Both work for me.

**Comment by blakeembrey** [comment]
> Just wondering what the main motivation behind that was

I'll be writing all new packages I work on as ESM. I don't want to develop solely for CommonJS anymore because that comes with more limitations, and I don't want to be maintaining both CJS and ESM because it becomes a confusing mess for documentation and supporting actual interesting features. For example, it's next to impossible to actually build a good ESM module that interacts with things in `import.meta` (e.g. `__dirname`, `__filename`, etc in CJS) unless you start doing other tricks too. Not to mention dependencies that I might want to use that are ESM only, which then also breaks anything else I publish as CJS. Easier to go ESM only for new versions as a maintainer, because you can always import CJS into ESM but not easily import ESM into CJS (it is, of course, possible, but it changes the API a ton because `import` is asynchronous, making it again impossible to build a complex package supporting both ESM and CJS).

> technically doesn't this remove support for NodeJS versions that don't support ES modules natively (less than Node 14.8-ish)?

v12.0.0 is the first version of node.js with ESM support. That was released 3 years ago. v12.17.0 was the first release without it being behind a flag (e.g. it just worked). v14.0.0 made it non-experimental (it stopped printing a message anymore). Maybe you're mixing this up with top-level await which come out in v14.8.0?

**Comment by blakeembrey** [comment]
More coherent documentation here too: https://github.com/sindresorhus/meta/discussions/15.

**Comment by mindplay-dk** [comment]
> @mindplay-dk Yes, I transpile my projects or use `ts-node/esm`. Both work for me.

With `ts-node/esm`, are you able to load both ESM and CommonJS packages?

Because I've [never seen that work](https://github.com/TypeStrong/ts-node/issues/1007#issuecomment-1118249190) - if you've managed to make that work, I would really like to learn how.

I'm completely onboard with your motivations for supporting ESM only, but the number of packages in the NPM ecosystem that actually ship with ESM versions still seem to be in the minority - going exclusively ESM *for dependencies* still seems pretty unrealistic for Node.JS projects, where practically everything depends on a CommonJS package... 😐

**Comment by blakeembrey** [comment]
It does work for me without anything special. I've been using it to develop https://github.com/borderless/site locally if you want an example. I'm pretty confident I have some ESM only (e.g. `@borderless/router`) and others are CommonJS only dependencies in there too.

**Comment by blakeembrey** [comment]
> the number of packages in the NPM ecosystem that actually ship with ESM versions still seem to be in the minority

Yep, but thankfully that doesn't matter for ESM imports - it "just works". The trouble is with the inverse, using ESM from CommonJS. Since that's such as huge pain, and ESM isn't doing away and is preferable for front-end and bundlers, it's easier for me (as a maintainer) to start migrating to ESM.

**Comment by mindplay-dk** [comment]
I did actually get my `ts-node` project up and going finally.

I had to change all my `import` statements to use `.js` for `.ts` modules - very confusing 🙁 but the TS team refuses to allow the `.ts` extension for `.ts` modules for some reason, omitting the file extension isn't supported by `ts-node` (yet) and apparently there's no way around it for now.

The other big realization for me was that CommonJS support in Node.JS itself is [kind of crippled](https://nodejs.org/dist/latest-v18.x/docs/api/esm.html#interoperability-with-commonjs):

> Named exports **may be available**, provided by static analysis as a convenience for better ecosystem compatibility.

It turns out they're almost never available (there's an article that explains it better [here](https://simonplend.com/node-js-now-supports-named-imports-from-commonjs-modules-but-what-does-that-mean/)) and so in practice, this won't work:

    import { red } from "ansi-colors";

Instead you have to manually destructure:

    import colors from "ansi-colors";
    const { red } = colors;

Since that's the required syntax for CommonJS imports, it's hard for me to understand why something, somewhere, in this mess of tools and internal node workarounds, why doesn't *something* in there just rewrite those imports when it *knows* you're importing from a CommonJS module? 🤷‍♂️

Man, there really is no such thing as "just write JavaScript" - there are so many pitfalls and nasty implementation details leading to all these crazy limitations and inconsistencies, and you have to just very slowly learn all of these crazy history lessons to get anything done in practice. 🙄

I just want to build a web app with Node.JS and TS, it doesn't seem that far fetched, ha ha.

Honestly, it's no wonder so many folks just stick with crappy PHP for their hobby projects...

**Comment by mindplay-dk** [comment]
(sorry for randomly blowing off steam in this thread 😳)

**Comment by blakeembrey** [comment]
All good, I totally understand. I felt like that when I first wanted to add ESM support to `ts-node` and the APIs were awful (I hear they still aren't great, but I don't work on it anymore). 

It helped knowing a lot of this ahead of time because I had tried to look into ESM support for `ts-node` way back when it first announced, so I conveniently had some background on these gotchas 😄 Plus I advocated for typings in TypeScript for a long time to avoid mixing and synthesizing ES-style imports and TypeScript-style `import x = require('x')`. I feel like everything would have been much simpler without preemptively shipping ESM syntax on top of CommonJS, but people want what people want.

The `import default from 'common-js'` was a major contention issue originally too, I remember they just had too much trouble and also the semantics are just different. Especially since `module.exports` could be anything at all, including a string or an object. I just assume all CommonJS modules can only be default imported and it works for me. 

TL;DR: I just follow path of least resistance. I'm using ESM for new projects, only import default when using CommonJS, and you always need to include `.js` for ESM imports too.
